### ISSUE TITLE
As a researcher, I want to enter or upload the provenance for my file so that I can provide other researchers with additional information about the history of my data.

### ISSUE THREAD

**Issue by djbrooke** [IQSS/dataverse#4343]
The mockups supporting this user story are available here:

https://drive.google.com/file/d/1AO_L0utJwjHEBxpj5b2xNiVkOjOG545D/view

This story covers the submission and validation of a JSON file, and the submission of manually entered provenance.

**Comment by pdurbin** [comment]
In cc09abf I stubbed out the validation that the prov file is JSON and can be parsed an a JSON object rather than a JSON array.

**Comment by dlmurphy** [comment]
In our usability testing of our Provenance mockups yesterday, we had one major piece of UI feedback that applied to this mockup:

![prov2](https://user-images.githubusercontent.com/22962740/33785798-a25c86ee-dc34-11e7-980b-ef20dce75bc8.PNG)

When you upload a provenance file, the "Upload Provenance File" box is still displayed and active. This led the participant to think that it was possible to upload more than one provenance file. Because he is very familiar with provenance, he was startled to think that we'd messed this up and allowed multiple file upload. 

There are two easy ways to fix this issue:
A. simply remove the file upload box once a provenance file has been uploaded, and only display it again if the user clicks the "X Remove" button.

OR

B. Once the user uploads a provenance file, the "+ Upload Provenance File" button label could be changed to "+ Replace Provenance File".

**Comment by pdurbin** [comment]
As of 7948579 we are storing both types of provenance data:

- "Bundle File" ([PROV-JSON](https://www.w3.org/Submission/prov-json/), a W3C standard): stored as an "auxillary" file for a datafile (like thumbnails)
- "Description" ("free-form", a text box): stored to a new `provfreeform` column on the `filemetadata` table

To be clear, this is where the two types will go after you click "Save Changes". There's still staging work to do for the UI. Clicking "Cancel" means nothing that's been staged will be persisted at all.

**Comment by dlmurphy** [comment]
Usability testing yesterday also got us feedback on some of the instructional text for this popup:

![prov1](https://user-images.githubusercontent.com/22962740/33786030-85a46106-dc35-11e7-96ab-1788d348755a.PNG)

Based on our participant's feedback, I recommend changing this message:

"Please add any information required to accurately document the history of this data file, including how it was created and how it was transformed."

TO

"In addition to a bundle file, or in place of one, you may add any information required to accurately document the history of your data file, including how it was created and how it was transformed."

I also recommend changing **Description** to **Provenance Description**.

**Comment by dlmurphy** [comment]
Lastly, @mheppler pointed out during usability testing that we need a "Cancel" button added to this popup modal:
![prov3](https://user-images.githubusercontent.com/22962740/33787551-bc660ec8-dc3b-11e7-8041-9db92fa5e260.PNG)

**Comment by mheppler** [comment]
What are we building?

File Upload Pg
- Edit button for each file with dropdown menu for Provenance (combine "Edit Tags" button as a dropdown option, as well as Delete, Restrict and Unrestrict)
- New Edit dropdown button in the table header, consistent with the Edit File dropdown button in the files table header on the dataset pg (the Provenance option will not be linked from this button, we are not supporting bulk/multi-file editing for Provenance)

Provenance (Edit) Popup
- New Provenance popup will provide help text and form input for the "Provenance Bundle" and the "Provenance Description", along with Save Changes and Cancel buttons
- Provenance Bundle form input is a file upload widget that validates JSON format (a nice to have is to validate against the W3C schema)
    - Upon uploading valid bundle, the upload widget is hidden, and a Remove and Preview button are displayed next to the bundle file name
    - Remove button removes uploaded bundle file, the upload widget is then displayed again
    - Preview button opens Provenance graph in new window
- Provenance Description form input is a textarea component
- Save Changes stages changes as part of the File Upload workflow
- Cancel deletes changes

Provenance (Confirmation) Popup
- Warning text explaining user must preview the Provenance info they added before publish, because you can not change it after you publish your dataset
- Continue button brings user back to File Upload workflow
- Cancel button brings user back to Provenance (Edit) popup

**Comment by TaniaSchlatter** [comment]
When we have a better sense of the graph, it may make more sense in terms of streamlining the flow to include the Preview in a modal.

**Comment by mheppler** [comment]
Mockup of the Edit dropdown button in the table header for File Upload pg, as referenced in checklist above.

This is intended to bring the functionality of the same Edit Files dropdown button on the dataset pg to the File Upload pg.

![provenance-fileupload-editselected-v3](https://user-images.githubusercontent.com/687227/33896376-a853f31c-df30-11e7-844d-f25a5e49b145.png)

**Comment by matthew-a-dunlap** [comment]
In 4378 we have discussed

**Comment by matthew-a-dunlap** [comment]
Sorry, meant to delete a comment not close the story!

**Comment by matthew-a-dunlap** [comment]
Questions: 

The Provenance (Confirmation) Popup happens when clicking "Save Changes" on the file upload page? Or does it happen when clicking save changes on the provenance (Upload) popup?
The notes on this popup says "explaining user must preview the Provenance". Are we enforcing this? In other words, are we requiring they view the graph and only allowing publication if that has happened?

**Comment by matthew-a-dunlap** [comment]
One more:

- On the file upload page, do we want to change the buttons for the multi-file restrict and delete? Not sure if it needs it but it may be good to re-align it with the new per-file dropdown.

**Comment by mheppler** [comment]
@matthew-a-dunlap 

- Confirmation popup is after save on the Provenance popup, before the user is returned to the File Upload pg
- There is no enforcing of a preview
- Yes, we want consistency with all our Edit buttons as scalable solutions going forward for all the actions related to files

**Comment by matthew-a-dunlap** [comment]
High level functionality To-Do:
[ ] Freeform provenance commands: read, delete, update
[ ] JSON-based provenance commands: read, delete, update
[ ] Wire up file page for provenance editing: option in dropdown and popups
[ ] Wire up upload page for provenance editing: updating file table with per file dropdowns, edit global dropdown and and popups
[ ] General popups work, independent of page shown: file upload and flow between the two popups
[ ] API access to added provenance commands

**Comment by mheppler** [comment]
Cleaned up the layout of the new Edit button in the table header on File Upload pg.

NOTE: Discovered that the file upload does not work. Once you select a file from your computer, it does not display in the file table, and there are no errors logged by glassfish. I assume that @matthew-a-dunlap is aware of this, based on the lengthy to-do checklist in his comment above.

Also, merged `develop` branch and resolved conflicts in `structure.css` in order to get bug fixes from @sekmiller in the work on #4051.

**Comment by dlmurphy** [comment]
I've finished the parts of the Provenance user guide that I can write from just the mockups. Once this reaches a point where I can demo the feature, I'll want to go back and make a few additions.

**Comment by matthew-a-dunlap** [comment]
If a user has uploaded provenance data to a file the past and goes to look at provenance for the file, how should that be represented and how should the popup look? 

If its an uploaded provenance file, should it be shown similar to https://github.com/IQSS/dataverse/issues/4343#issuecomment-350377415? Does there need to be some difference to communicate that this file is already there?

If a user goes to this popup and doesn't input any changes, should we still say "save changes" (which may confuse the user that they have made changes that they haven't)?

**Comment by TaniaSchlatter** [comment]
Hi Mathew,

Comments inline below:

> On Jan 9, 2018, at 7:41 PM, matthew-a-dunlap <notifications@github.com> wrote:
> 
> If a user has uploaded provenance data to a file the past and goes to look at provenance for the file, how should that be represented and how should the popup look?
> 
> If its an uploaded provenance file, should it be shown similar to #4343 (comment)? Does there need to be some difference to communicate that this file is already there?
> 

It should look like the mock-up in the comment. It shows a provenance file (test-prov.json) has been uploaded and it there. The upload box should not show, however, as per Derek’s comment. 

> If a user goes to this popup and doesn't input any changes, should we still say "save changes" (which may confuse the user that they have made changes that they haven't)?
> 
It should say “save changes” as shown in the mock-up in the comment. The user can cancel if they don’t put in any changes.

Thanks!

- Tania

**Comment by matthew-a-dunlap** [comment]
@TaniaSchlatter The problem that I see is that there are two different states:

1. A user has uploaded a provenance file but not saved it
2. A user previously uploaded a provenance filed, saved it, and is now looking at the file through this screen 

Is our intention to have the same visual state (as shown in the mockup) for both? Won't our users be confused if they upload the file, click save, reopen the popup to be sure and see the same UI as before?

**Comment by TaniaSchlatter** [comment]
The same visual state works for both. We could say “save” on the button in both rather than “save changes,” but I don’t think it is necessary, and would not want to do that without checking for consistency of wording on save-type buttons in general.

- Tania

**Comment by matthew-a-dunlap** [comment]
I have updated my checklist above with the state of this story. All the main functionality works and is wired up. The confirm popup is not committed but mostly done on my local branch. Hopefully will be able to wrap this up on Tuesday.

**Comment by mheppler** [comment]
@matthew-a-dunlap in your to-do above, you ask, _What should happen if either form of provenance data fails to save?_

I suggest using an error message from the bundle, like we do for Solr, that reads something like, _Due to an internal error, your provenance information was not saved_.

**Comment by matthew-a-dunlap** [comment]
My most recent commit was not tagged correctly to show up here, but was committed to the branch ( https://github.com/IQSS/dataverse/compare/4343-prov-upload-valid ). The confirm popup is now functional.

I have also updated the list above with a few more bugs/issues that need to be worked on. None of them look major at all.

**Comment by mheppler** [comment]
Cleaned up UI of Provenance workflow on File Upload pg. Checked off the following to-do items:
- Provenance main popup aesthetics
- Provenance confirm popup aesthetics

Was not able to replicate the following bug:
- If provenance popup is closed, cannot be reopened without refreshing page

**Comment by matthew-a-dunlap** [comment]
This branch is ready for code review. I decided not to expose the provenance get commands via api at this point as I think there is still some flux in that space (though likely it would be just returning the json file each time). I have added a note to the relevant provenance story.

**Comment by matthew-a-dunlap** [comment]
Derek was looking this over for documentation and noticed an issue where the provenance data will not save correctly when editing is accessed from the file upload and create dataset pages. I had lost track of this need when programming, so I am pulling this back into development.

This also made me realize another issue: on the edit files page, it is confusing what all the "Save Changes" buttons are doing. In the provenance popup, should clicking "save changes" here save the provenance data to the database?
![screen shot 2018-01-19 at 5 44 46 pm](https://user-images.githubusercontent.com/8901015/35175305-06fa68e6-fd41-11e7-8340-3bd4e98f91cc.png)

or should it just stage the changes until save changes it then presssed again here?
![screen shot 2018-01-19 at 5 44 31 pm](https://user-images.githubusercontent.com/8901015/35175332-332ac960-fd41-11e7-846a-6d91f5b1162b.png)

If we want the provenance data to not be committed until the Edit Files Page save changes button is clicked, it still feels to me to be a bit ambiguous.

**Comment by matthew-a-dunlap** [comment]
This branch alllllmost ready for code review. I need to tweak how one of our commands is called and do code cleanup. Otherwise its good.

I deployed the current state to https://apitest.dataverse.org , which is close enough to be looked over @dlmurphy @mheppler @TaniaSchlatter

**Comment by matthew-a-dunlap** [comment]
Had a discussion with the design team this morning. They've identified the need to add the popup to the file page. This is not as simple as just wiring the popup up for another page, as the popup in this page needs to trigger saving within itself. Its not that much work in the end, but I bet it won't be done today.

I pointed out a design issue I saw, that the saving of provenance changes is inconsistent depending on the page. In the file page saving happens in the popup while on the other pages users can complete the popup and leave without their work being saved. I also pointed out that our approach of wiring up these interactions in many places introduces complexity to the codebase which overall can impact the ability to work with the code.

**Comment by matthew-a-dunlap** [comment]
This branch is complete besides one minor issue due to how we calculate DataFile equality. We check if DataFile.id is equal, but because DataFiles can have null as their id this check will return true for different DataFiles. I could work around it but I think it is best to talk to @scolapasta on Monday if we could expand on the equals method.

**Comment by mheppler** [comment]
Comment by @jacksonokuhn in #4381
> On the upload page, can we refer to the provenance as "Documents" rather than "Bundles"? We're storing the data as bundles on the backend, but they're gonna be uploaded as standard PROV-JSON docs. The current wording may confuse people.

**Comment by dlmurphy** [comment]
After discussing with Jackson, we're going with "Provenance File" as our terminology. I've updated the UI to reflect this, and I'm currently updating the documentation.

**Comment by matthew-a-dunlap** [comment]
There are a few "outstanding issues" related to the popup work:
- Saving provenance data should only be possible if the file in question did not have any in previous published version. Currently one could go and change the provenance after publication.
- The provenance popup rendering is not tied to any sort of set-up variable to inform it whether the prov system exists for the installation. There are open questions about whether the prov system will be more required or optional.

These may require pulling the popup code back into development, but I think the code in https://github.com/IQSS/dataverse/pull/4422 should be reviewed as-is for other issues first.

**Comment by matthew-a-dunlap** [comment]
Had discussion with @scolapasta & @ekraffmiller around equality and entity objects today. Tomorrow I plan to move forward on those and try to get this story _really_ into code review.

**Comment by matthew-a-dunlap** [comment]
I have rewritten functionality for the provenance popup to better handle issues around testing equality of datafiles. Also, provenance dropdown options and popup will only show up if a provenance address has been set via rest call. I plan to ask around tomorrow about how best to check if a datafile in a previous version had provenance and tie that into render logic.

I have turned off rendering of dropdown elements for tags/restrict/unrestrict/delete that were added to the editDataFiles.xhml on the branch. The work to make these functional is non-trivial, especially due to the sprawling nature of various file pages. Even with restrict/unrestrict/delete, where there are already method to an action based on selected checkboxes (and we are looking to allow a user to directly act on one file) I still feel this work should be handled in its own story. We have discussed that for #3488 we may want to refactor the tags functions, and a similar discussion should be had for restrict/unrestrict/delete.

**Comment by matthew-a-dunlap** [comment]
This is the current look of the provenance popup after the provenance json has been sent to CPL and can no longer be edited. If anything, we should probably use different wording for the freeform string @dlmurphy @mheppler 

![screen shot 2018-02-08 at 1 04 47 pm](https://user-images.githubusercontent.com/8901015/35990049-d5a17786-0cd0-11e8-851a-8b1777800ee7.png)

**Comment by dlmurphy** [comment]
@mheppler and I have put together some screenshots depicting new messaging for the provenance popup, along with what it should look like in the use case @matthew-a-dunlap just mentioned. Please excuse any minor layout weirdness in these mockups, the only real difference here is the text.

![prov_popup_1](https://user-images.githubusercontent.com/22962740/36041178-b84426f0-0d95-11e8-98ed-7ff5d281b4de.PNG)

<img width="605" alt="prov_popup_2" src="https://user-images.githubusercontent.com/22962740/36041182-bb82eac2-0d95-11e8-8e34-36876d69259d.png">

![prov_popup_3](https://user-images.githubusercontent.com/22962740/36041188-bea6bb66-0d95-11e8-97be-c4e53f492467.PNG)

**Comment by dlmurphy** [comment]
Just collaborated with @TaniaSchlatter to rewrite the text in that yellow warning popup. Here's a mockup:

![prov_popup_4](https://user-images.githubusercontent.com/22962740/36053291-3094b8d0-0dbf-11e8-985b-76782c445828.PNG)

And here's the text for easy copy/paste:

Once you publish this dataset, your provenance file can not be edited or replaced. 

Select "Cancel" to return the previous page, where you can preview your provenance file to confirm it is correct.

**Comment by jacksonokuhn** [comment]
I'm realizing that we need some more text on the upload screen :/ Basically we need to tell users that the title they give their datafile should be the same as the title it has in the prov they're uploading. The other option (which is way more of a pain in the butt) is to actually add another input box where they can give us the name of their datafile in the uploaded provenance. We need this in order to connect the prov getting uploaded to the prov in our system.

**Comment by matthew-a-dunlap** [comment]
@jacksonokuhn We can get the name of their datafile from the db and pass it in the rest call when we send the provenance to CPL. Is there some reason we need the user to specify it instead?

**Comment by sekmiller** [comment]
@jacksonokuhn - instead of relying on the name being the same, which could be problematic across datasets and installations, can we use the file DOI/PID?

**Comment by jacksonokuhn** [comment]
either way works actually, so long as we make it clear to the user

**Comment by matthew-a-dunlap** [comment]
This is being pulled back into development as we need the user to provide us extra information for use when sending their upload prov.json to CPL. Specifically, we are unable to tell in many cases which file mentioned in the provenance json relates to the DataFile we are sending provenance for. This can often be inferred by comparing file names but if names have been changed for some reason the user should still be able to work with CPL.

To solve this, when the user uploads their provenance json we will mine the filenames out of the json and use them to help the user tell us which file relates to the DataFile they are on. This interaction will happen under the file upload dialog in the main popup. At a later date we may also show the user a preview of their provenance graph to assist them, but this enhancement can be done during/after work to interface with the visualization library in #4346.

There are some outstanding questions on this work:
* The number of file options presented to the user may be very large, too large for a dropdown. How are we going to help the user with this selection? An autocomplete text box has been suggested.
* If the names match it would be best to present that option to the user first. Is this a hard requirement? If so, how do we do this alongside the selection process above.
* UI interaction is one thing, but how do we do this via command line? The easiest option is to require API users to provide the correct name along with their json in the call.

**Comment by matthew-a-dunlap** [comment]
Starting a new ToDo list to tackle the json parsing, dropdown work, and feedback from code review before this was pulled back into develop:
- [ ] Don't use the "get provenance" commands in the popup, do the getting directly.

**Comment by matthew-a-dunlap** [comment]
The current state of affairs (before my long weekend): No code has been started for this story but after discussing the need in our meeting yesterday I investigated the use of an autocomplete dropdown for the task. The code for one in permissions copy-pasted into the popup fragment cleanly and is a good starting point. The next step is understand the w3c provenance spec a tiny bit better and parse the json. The basic need is to parse all the sub-tags to the activity tag and present those as options using the name tag inside each sub-tag if it exists ("*:name" to be specific, for example "rdt:name"). I was also going to check to see if the DataFile Name matches the sub-tag name and present it first in the suggestion list.

There is still an open question of what to do if two tags share the same name. My idea for an approach was present them both as options, but maybe try to parse out some other identifying piece of info. Doing this in a way that is useful generically is likely challenging. Aside from this edge case I do not see this work as taking that much time, tho I have not had to do much json parsing in java (I have done xml parsing and remember dealing with those libraries to be more painful than I expected).

**Comment by matthew-a-dunlap** [comment]
@mheppler This is ready for you to look over. There is remaining work, but its either pretty backend or API. The one exception to this is the UI issue related to upload => remove => upload again. I'm pretty stumped on that one.

**Comment by matthew-a-dunlap** [comment]
![screen shot 2018-02-27 at 12 42 24 pm](https://user-images.githubusercontent.com/8901015/36744829-cc686f7c-1bbb-11e8-983e-b318b5c4787a.png)
@mheppler @dlmurphy Here is the error dialog that happens when there is an error during uploading and parsing of the provenance json. Let me know what you think of the wording. We will also need another error (or warning?) for when the provenance uploaded by the user does not have any entities (files) that can be parsed out for the user to select as connected to their DataFile.

**Comment by matthew-a-dunlap** [comment]
![screen shot 2018-02-27 at 1 46 09 pm](https://user-images.githubusercontent.com/8901015/36748729-b6ce3042-1bc6-11e8-9490-3d74288b86d4.png)
Here is a sample of the provenance autocomplete for selecting the entity @dlmurphy @jacksonokuhn

**Comment by matthew-a-dunlap** [comment]
All that looks to be left is fixing the IT tests and maybe writing another test or two for the parser code!

**Comment by matthew-a-dunlap** [comment]
Deployment notes:
- Run db updates (4.8.5 to 4.9)
- Make sure this system setting is in place (The documentation is is #4364). curl -X PUT -d 'http://localhost:7777' http://localhost:8080/api/admin/settings/:ProvServiceUrl
   - The url contents do not matter for this branch, just its existence is needed to enable provenance

Testing notes:
- Provenance popup can be accessed from these locations: Dataset creation, file upload, edit files (off main page), file page. These all need to be tested, especially file page as its flow is the most different.
- Its possible there is an edge case I've missed when adding and removing the prov json from the popup. Test adding, removing and readding while the popup is up, or adding, removing, closing and reopening. Etc etc.
- You will need prov json examples with entities in them. I pulled mine from https://github.com/CamFlow/cytoscape.js-prov/tree/master/json (elisabeth & example look good)

**Comment by dlmurphy** [comment]
Noticed a bug while I was poking at the test server. When using the entity connector dropdown on the add provenance window from the create dataset page, I used the scroll wheel on my mouse, and it caused the dropdown to move up and down on the screen while everything else stayed static:
![scrollbug](https://user-images.githubusercontent.com/22962740/37062866-abac63f6-2165-11e8-9d7d-cdbb20aaf014.PNG)

I was using Chrome. Spoke with Matthew about it and he said he's seeing this bug on Chrome and Safari as well.

**Comment by matthew-a-dunlap** [comment]
Thanks to @mheppler for the info on fixing the dropdown scroll. I fixed autocomplete dropdowns in these popups as they were all exhibiting this behaviour and js fix is pretty benign:
Also touched autocomplete dropdowns in:
- Harvest Clients (Destination Dataverse)
- Manage Groups (Role Asignee Name)
- Dataset Permissions Manage Files (User/ Group Name Assign)
- Role Assign (User / Group Role Assign)

**Comment by matthew-a-dunlap** [comment]
Review fixes:

- [ ] Store cplId in the same format in the database as it is in the code (Pete).
- [ ] Check if provenance is enabled before running its saving function on DatasetPage & EditDataFilesPage

**Comment by matthew-a-dunlap** [comment]
Note: With the review fixed done so far, a new system variable is needed to enable provenance, ProvCollectionEnabled . It is a boolean. We may want to have it enabled by default.

**Comment by matthew-a-dunlap** [comment]
Prov is in a pretty good place as I leave for a few days. Aside from the new messaging, there are only a few outstanding bugs, mostly related to the prov popup not being updated correctly by jsf at certain points. See the checklist above for more info.

In terms of confirm dialog messaging, folks should be able to walk through the flows and see the outputted sentence structure. The ordering almost certainly needs changing alongside the words themselves.

**Comment by mheppler** [comment]
Reviewing the latest and greatest from @matthew-a-dunlap in this branch on the apitest server, and I found a few things of note:

- [ ] Edit Files pg > no success msg about "Provenance information will be added to this version of this dataset once you click on the Save Changes button" after clicking Continue btn in Provenance popup, resulting in no visual indicator to the depositor that anything happened when the popup closed
- [ ] Provenance popup > maybe some of the warning text msg about "Your Provenance File will become permanent once you click Save Changes" from the confirmation popup could be displayed above the upload component when the prov file is added to an existing published file, just like the "This Provenance File has been published and cannot be replaced or removed" msg is displayed
- [ ] Provenance confirmation popup > fix display of msg text so it does not break to a second line after the warning icon, maybe move Preview button out of the bottom action btns and into the popup message body
- [ ] Versions tab > no transaction is recorded when creating a new draft by adding prov to existing file
- [ ] Version Differences Details popup > same as the Version tab, no transaction is recorded when creating a new draft by adding prov to existing file

**Comment by matthew-a-dunlap** [comment]
I've implemented the changes pointing out by @mheppler . There is text related to the new alert dialogs that should probably be reviewed @dlmurphy . Due to limitations of the File Page I went with the solution of only showing one alert dialog per popup save, even if both Provenance Description and Provenance File are saved. 
![screen shot 2018-04-23 at 7 56 43 pm](https://user-images.githubusercontent.com/8901015/39158927-bd4458bc-4730-11e8-9a07-28c6eb6abb15.png)

**Comment by matthew-a-dunlap** [comment]
Note: During discussion @mheppler and I discussed how a user on the editDataFiles page may go to upload their prov file and then think they need to save. That save is unneeded and will cause a draft to be created. At the time of writing we were of the opinion this should be dealt at a later date by preventing users from saving empty drafts.

Note 2: I fixed an outstanding bug with selected entity not being displayed correctly after changing other prov data for a DataFile.

**Comment by matthew-a-dunlap** [comment]
Deployment notes:
- Run db updates (4.8.5 to 4.9)
- Enable provenance : curl -X PUT -d 'true' http://localhost:8080/api/admin/settings/:ProvCollectionEnabled

[this post is being actively edited]

**Comment by matthew-a-dunlap** [comment]
Thanks again @kcondon for the digging. Here are a few notes on some of the issues uncovered. We can touch base Monday afternoon when I am back in the office:

* Regarding *on edit file metadata with prov disabled* **:** For me in chrome 66.0.3359.139 the multi-menu shows up correctly infront of the table, and has options for delete/restrict/unrestrict.
* Regarding *Using Chrome, cannot save a prov file* **:** I'm unable to reproduce this on my machine running chrome 66.0.3359.139 . Was this a consistent issue?
* Regarding *Prov popup does not open in Chrome if prov file previously saved* **:** I'm not able to reproduce this in Chrome 66.0.3359.139.
* With the two chrome issues above, even if I can't reproduce it these problems may point to the code being brittle if a plugin or something is making things slightly out of wack.
* Regarding *Dataset version details* **:** version details should not have any info on the prov file as only the description is versioned. 
* Regarding *Prov desc does not show up on file versions tab...* **:** Good catch, I forgot there was version info on that page!
* Regarding *While editing prov from edit file metadata page... continue but don't save dataset* **:** Good catch, I think I'm not clearing out my delete action correctly.
* Regarding various issues around saving without prov entity **:** Once the popup rendering issues are fixed the user should be unable to save without an entity for the prov JSON. I did not add an extra safety check on publish for it but can do so if needed. For what its worth I'm unable to get the popup to save without an entity


After we touch base and you are done with any checks for this round I'll create a todo checklist to capture the needed work.

**Comment by matthew-a-dunlap** [comment]
Here are the issues I've taken a stab at fixing. I don't want to direct edit @kcondon 's list as he could confirm my fixes:
- Prov desc does not show up on file versions tab...
-  Server log error on edit popup if access via edit file metadata...

**Comment by kcondon** [comment]
OK, tested the above fixes and they work.

**Comment by matthew-a-dunlap** [comment]
More fixes info:
- Fixed: "If prov file previously saved for unpublished draft"
- I think I fixed but couldn't quite reproduce: "While editing prov from edit file metadata page, if add prov file"
- Fixed: All API issues. As part of this I removed the DeleteProvFreeform API command as it was redundant/unneeded (update can be used for the same functionality). Also note that Prov JSON can be uploaded for a published DataFile as long as JSON doesn't already exist.

**Comment by kcondon** [comment]
All of the above are fixed. A minor note about maybe providing a curl example for create prov, showing --upload-file

**Comment by matthew-a-dunlap** [comment]
After discussion with @scolapasta and @sekmiller , it looks like one of the saving flows on the editDatasetPage is not saving with best practice and the fileMetadata is likely not in a useful state afterwards for the prov command to use and save. Its possible prov could work around this but it'd be pretty messy, I'm going to first take a look at what a SaveDataFileCommand would look like.

**Comment by matthew-a-dunlap** [comment]
After talking with @landreev , I was able to arrive at a fix that did not involve digging deep into saving commands. Prov changes now trigger the full dataset update in editDatafilePage, instead of the optimized but less robust flow.

When investigating this change, here are the things to look for:
- Saving prov json and prov freeform from editDatafilePage after a file upload (not editing prov during the upload action) should work.
   - Try doing this with a tabular file and tabular data tags, there was some change these could conflict with each other tho from my testing it looked ok
- Along with this I found a bug where editing just the entityId did not save. Should work now.
